### ISSUE TITLE
ref is called after componentDidUpdate?

### ISSUE THREAD

**Issue by stevenguh** [reactjs/react-modal#338]
### Summary:
ref is called after componentDidUpdate

### Steps to reproduce:

 1. Attach a ref in any component in the children of the modal
 2. The componentDidUpdate will get called before ref get attached to any children of the modal.

### Expected behavior:
ref should get attached before componentDidUpdate
### Additional notes:
I created a couple codepens to further explain this issue.

This [codepen](http://codepen.io/anon/pen/vxLVmY?editors=0111) will reproduce this bug. Simply open the console, and click the Trigger Modal button. You should see two log messages, "componentDidUpdate" and follow by a "attach ref".

This [codepen](http://codepen.io/anon/pen/MpKPoQ?editors=0011) will show a normal react lift cycle if I remove change the Modal component to a div. When you click Trigger Modal button, you should see two "attach ref" follow by a "componentDidUpdate" after clicking the Tigger Modal button.

This [codepen](http://codepen.io/anon/pen/KWVrKB?editors=0011) is a workaround I discovered. Open up the console, and click Tigger Modal, then click the number button. You should see two "attach ref" follow by a "componentDidUpdate" after clicking the number button.
Basically, if you can handle all the state and update in a component without declaring as the direct children of Modal, then you can create another component to wrap around all the logic, and put that component as the child of the Modal component.

I think this behavior would surprise a lot of people if they need the correct sequence of attaching ref and componentDidUpdate. I skimmed through the source a little bit, and I think it's likely relate to how ModalPortal get created. I think at the very least this needs to be either documented or fixed.

**Comment by MrRacoon** [comment]
Just got burned by this as well.

**Comment by micdah** [comment]
Yes this doesn't seem like the behavior one would expect, +1

**Comment by diasbruno** [comment]
@stevenguh @griffinmichl @MrRacoon @micdah and everyone. 

This is the "expected" behavior when you create a closed `Modal` (note that at this point the ModalPortal is not rendered, so no refs yet).

The lifecycle is actually correct. On the class `Example`, the refs will be built after the update, since the `ModalPortal` will be mounted after this update. Your `ModalContent` wrapper works as expected, because the refs are local to the this class.

Hope now it's explained why it works this way.

**Comment by diasbruno** [comment]
@outdooricon Oh my...That was the worst explanation 🤦‍♂️ .

Ref: [Adding a Ref to a DOM Element](https://reactjs.org/docs/refs-and-the-dom.html#adding-a-ref-to-a-dom-element)

[ModalPortal.js#L301](https://github.com/reactjs/react-modal/blob/master/src/components/ModalPortal.js#L301)

This line will prevent any children from been mounted on the react tree, so if the modal is closed the `ref` callback won't be called, only when the modal is open.

**Comment by erin-doyle** [comment]
I'm still really unclear how to solve this.  I want the close button to be focused as soon as the modal opens.  Can someone provide an example as to how to accomplish this?  I've tried calling`handleOpenModal()` and that still seems to happen before my `ref` has been set.

**Comment by LENGUYENMINH** [comment]
Totally agreed @stevenguh. I had the same problem. 

Issue : ref in componentDidUpdate is ref of element of "render before", not the "updated render". It's a bug

**Comment by ysz** [comment]
+1

@erin-doyle any update?

**Comment by erin-doyle** [comment]
@ysz this is not my library so no update from me.  IIRC in the project where I was using this I ended up having to set: 
`shouldFocusAfterRender={true}`
`shouldReturnFocusAfterClose={false}`

I gave up trying to set focus to anything specific within the modal when it opens and had to add code managing focus explicitly on `onRequestClose` to make sure the focus returns to the corrrect location when the modal closes.
